cfgov/archival/utils.py

```diff
@@ -0,0 +1,118 @@
+import json
+import logging
+
+from django.apps import apps
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db.migrations.recorder import MigrationRecorder
+from django.utils import timezone
+
+from wagtail.coreutils import find_available_slug
+from wagtail.models import get_streamfield_names
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_last_migration(app):
+    Migration = MigrationRecorder.Migration
+    app_migrations = Migration.objects.filter(app=app).order_by("-applied")
+    last_migration = app_migrations.first()
+
+    if last_migration is None:
+        return ""
+
+    return last_migration.name
+
+
+def export_page(page):
+    """Get a copy of the page as JSON."""
+
+    # This includes its app_label, model, and the latest migration applied for
+    # the app, allowing any import to quickly verify whether import is
+    # possible (due to app or model availability and/or schema changes).
+    page_export = {
+        "app_label": page.content_type.app_label,
+        "model": page.content_type.model,
+        "last_migration": get_last_migration(page.content_type.app_label),
+        "exported_at": timezone.now().isoformat(),
+    }
+
+    # The serializable_data comes from django-modelcluster.
+    page_export["data"] = page.serializable_data()
+
+    # Convert streamfields, which serialize as strings containing JSON, back
+    # into objects. This will keep the final JSON cleaner, and not with
+    # JSON-in-a-string-in-JSON
+    streamfield_names = get_streamfield_names(page.specific.__class__)
+    for streamfield_name in streamfield_names:
+        page_export["data"][streamfield_name] = json.loads(
+            page_export["data"][streamfield_name]
+        )
+
+    # Strip the pk out of the data so that, on import, from_serializable_data()
+    # always creates a new object.
+    page_export["data"]["pk"] = None
+
+    # Dump the page data to JSON
+    page_json = json.dumps(
+        page_export, ensure_ascii=False, indent=4, cls=DjangoJSONEncoder
+    )
+
+    logger.info(f"Exported {page.slug} to JSON")
+    return page_json
+
+
+def import_page(parent_page, page_json, slug=None):
+    page_data = json.loads(page_json)
+
+    # Verify that we can load import the page
+    app_label = page_data["app_label"]
+    model_name = page_data["model"]
+    page_last_migration = page_data["last_migration"]
+
+    # Get the specific model that the imported page belongs to
+    try:
+        model = apps.get_model(app_label, model_name)
+    except LookupError as exc:
+        logger.error(
+            f"Unable to import page of type {app_label}.{model_name}; {exc}"
+        )
+        raise
+
+    # Compare the last migration
```

Comment:
Isn't [squashing](https://docs.djangoproject.com/en/5.1/topics/migrations/#migration-squashing) a big problem here? If an export is made with a migration that is subsequently squashed away (which ideally happens frequently), it'll never be importable.

Reply:
I kind of want it to be a problem? There's no way we can guarantee data is importable after a squash, or after a new migration — it might be, but IMHO it's better to not attempt and to bail. So that's what this does.

Reply:
If I'm reading this right, having the `import_page` method raise exceptions on failure will result in an unhelpful 500 error page to the user, even in non-exceptional cases like the mismatched migration one. Is there a way to use the Django messages framework to have these errors displayed to the user so they have the knowledge needed to resolve them?

Reply:
I've used `form.add_error()` to that field in a `try`/`except` in the view instead, to scope the error to the upload form field for the user. It looks like this:
<img width="878" alt="image" src="https://github.com/user-attachments/assets/b6dca2c8-62f6-40b7-aaf1-9a142ffdd2f6">